### PR TITLE
Added player voted event

### PR DESCRIPTION
**Pull request by Mr-VIT**
Added:
- `PlayerVotedEvent` in Server 
- `IPlayerVotedEvent` in API
- `OnPlayerVoted` Impostor.Plugins.Example

`RpcCalls.CastVote` is not used because Host player never calls it.

ToDo:
- [x] keep original NetId for disconnected players

### CODE REVIEW

**Review by reecebenson — APPROVED** [review]

**Comment by kantesilvestrs** [comment]
I implemented this via
``` csharp
[EventListener]
public async void OnPlayerVotedEvent(Impostor.Api.Events.Player.IPlayerVotedEvent e)
{
     var whoVoted = e.PlayerControl.PlayerInfo.PlayerName;
     var votedFor = e.VotedFor.PlayerInfo.PlayerName;
     await Bot.SendMessage($"{whoVoted} voted for {votedFor} in {e.Game.Code.Code} lobby");
}
```
and got this error when a person voted skip
```
[19:01:41 WRN] deadBodyPlayerId was byte.MaxValue
[19:01:41 WRN] deadBodyPlayerId was byte.MaxValue
Unhandled exception. System.NullReferenceException: Object reference not set to an instance of an object.
   at Impostor.Plugins.ImpostorCord.Handlers.GameEventListener.OnPlayerVotedEvent(IPlayerVotedEvent e)
   at System.Threading.Tasks.Task.<>c.<ThrowAsync>b__140_1(Object state)
   at System.Threading.QueueUserWorkItemCallbackDefaultContext.Execute()
   at System.Threading.ThreadPoolWorkQueue.Dispatch()
```

File: src/Impostor.Api/Events/Game/Player/IPlayerVotedEvent.cs [hunk 1]
```diff
@@ -0,0 +1,12 @@
+using Impostor.Api.Net.Inner.Objects;
+
+namespace Impostor.Api.Events.Player
+{
+    public interface IPlayerVotedEvent : IPlayerEvent
+    {
+        /// <summary>
+        ///     Get the player he voted for.
+        /// </summary>
+        IInnerPlayerControl? VotedFor { get; }
```
Comment (kantesilvestrs) [inline src/Impostor.Api/Events/Game/Player/IPlayerVotedEvent.cs:10]: This is not warning my IDE that VoteFor might be null and allows me to compile

Comment (Mr-VIT) [reply src/Impostor.Api/Events/Game/Player/IPlayerVotedEvent.cs:10]: check example pls https://github.com/Impostor/Impostor/pull/181/commits/6d02e86856ddadd9545f302bd8f794c0c17d5d3f

Comment (Mr-VIT) [reply src/Impostor.Api/Events/Game/Player/IPlayerVotedEvent.cs:10]: add `#nullable enable`

Comment (kantesilvestrs) [reply src/Impostor.Api/Events/Game/Player/IPlayerVotedEvent.cs:10]: My bad - I forgot visual studio doesn't enforce nullable `?` when typing nullables

Comment (AlexejheroYTB) [inline src/Impostor.Api/Events/Game/Player/IPlayerVotedEvent.cs:10]: Add VoteType enum and property: Vote, Skip, or None

If VoteType is skip or none, VotedFor is null.

_(alternative enum name is good as well)_

File: src/Impostor.Server/Net/Inner/Objects/InnerMeetingHud.cs [hunk 2]
```diff
@@ -168,6 +168,21 @@ public override void Deserialize(IClientPlayer sender, IClientPlayer? target, IM
                     if ((num & 1 << i) != 0)
                     {
                         _playerStates[i].Deserialize(reader);
+                        var playerState = _playerStates[i];
+
+                        if (playerState.DidVote)
```
Comment (AlexejheroYTB) [inline src/Impostor.Server/Net/Inner/Objects/InnerMeetingHud.cs:173]: Also trigger the event if the player didn't vote, and set VoteType to None.

**Review by AlexejheroYTB — CHANGES_REQUESTED** [review]

File: src/Impostor.Server/Net/Inner/Objects/InnerMeetingHud.cs (hunk 2 above)
Comment (realfaceitkantesilvestrs) [reply src/Impostor.Server/Net/Inner/Objects/InnerMeetingHud.cs:173]: This would allow plugins to avoid writing logic to figure out who didn't vote

**Comment by AeonLucid** [comment]
I don't like this pull request. The CallAsync of the eventmanaged is not awaited and should be moved to the `CastVote` RpcCall handler.

**Comment by realfaceitkantesilvestrs** [comment]
Just tested with different people connecting to the server - vote count and voted for are correct but people who voted are completely incorrect on every single instance
```
Voting for BBTGAU-2a8a28cf-1a28-466d-b7fc-6fca2f3ef802
List of all voting sessions

Voting Session - 1
Players alive: Odini,k1ro,SwagY,McQuail,YgKey,LucidKujo,Nightfur,jay,eskay,Limur
eskay,k1ro,YgKey,jay,Limur,Odini,LucidKujo,SwagY,McQuail voted skip
Nightfur voted LucidKujo

Voting Session - 2
Players alive: Odini,k1ro,SwagY,McQuail,YgKey,LucidKujo,jay,eskay,Limur
eskay,McQuail,jay,SwagY,k1ro,Limur voted skip
Odini voted YgKey
Nightfur voted Odini

Voting Session - 3
Players alive: Odini,k1ro,SwagY,McQuail,YgKey,LucidKujo,jay,Limur
McQuail,Limur,SwagY,eskay,LucidKujo,k1ro voted Limur
Odini voted LucidKujo
jay voted skip

Voting Session - 4
Players alive: Odini,k1ro,SwagY,McQuail,YgKey,LucidKujo
McQuail,eskay voted YgKey
SwagY,Limur,Odini,Odini voted LucidKujo
k1ro voted skip

Voting Session - 5
Players alive: Odini,SwagY,McQuail,YgKey
eskay voted YgKey
SwagY,k1ro voted Odini
```

**Comment by Mr-VIT** [comment]
@AeonLucid:
> The CallAsync of the EventManager is not awaited

There is any reason to make `InnerMeetingHud.Deserialize` async and await a non-canceled event?


**AeonLucid**:
>  should be moved to the `CastVote` RpcCall handler.

I already answered: 
> RpcCalls.CastVote is not used because Host player never calls it.

There is no other way to catch Vote from Host player

**AeonLucid**:
> I have reverted the `DidVote` change because this is how Among Us parses it, meaning, so should we.

Why? The Event now fires at the end of each meeting for every player who has not voted as voted for playerId `14`
Or should I check playerId right before calling the event? Then why would `DidVote` be `true` if the player didn't vote?

**Comment by AeonLucid** [comment]
@Mr-VIT 
> There is any reason to make InnerMeetingHud.Deserialize async and await a non-canceled event?
Yes, this will cause threading issues.

File: src/Impostor.Server/Net/Inner/Objects/InnerMeetingHud.cs [hunk 3]
```diff
@@ -176,11 +177,27 @@ public override void Deserialize(IClientPlayer sender, IClientPlayer? target, IM
...
+                                        voteType = VoteType.None;
+                                        break;
+
+                                    default:
+                                        voteType = VoteType.Player;
+                                        VotedForPlayer = _game.GameNet.GameData.GetPlayerById((byte)playerState.VotedFor)?.Controller;
+                                        break;
+                                }
+
+                                _eventManager.CallAsync(new PlayerVotedEvent(_game, clientPlayer, player.Controller, voteType, VotedForPlayer));
```
Comment (Mr-VIT) [inline src/Impostor.Server/Net/Inner/Objects/InnerMeetingHud.cs:200]: ToDo?:
 - [ ] **await** `_eventManager.CallAsync(new Event)`

**Comment by realfaceitkantesilvestrs** [comment]
@Mr-VIT 

I just checked my logs and talked with people:

Here is what events picked up
```
Voting Session - 2
Players alive: Odini,k1ro,SwagY,McQuail,YgKey,LucidKujo,jay,eskay,Limur
eskay,McQuail,jay,SwagY,k1ro,Limur voted skip
Odini voted YgKey
Nightfur voted Odini
```

And here is how they actually voted
```
Voting Session - 2
Players alive: Odini,k1ro,SwagY,McQuail,YgKey,LucidKujo,jay,eskay,Limur
McQuail,jay,SwagY,k1ro,Limur,Odini voted skip
Nightfur voted YgKey
eskay voted Odini
```

So it seems that the target is correct, but voters are moved one step forward so when `Odini voted YgKey` and `Nightfur voted Odini` it was actually `Nightfur voted Ygkey`

And what is even stranger that here 
```
Voting Session - 5
Players alive: Odini,SwagY,McQuail,YgKey
eskay voted YgKey
SwagY,k1ro voted Odini
```

`eskay` was dead and `Odini` voted `YgKey`

**Comment by Mr-VIT** [comment]
@realfaceitkantesilvestrs looks like host can send data of players in "wrong" sequence.
https://github.com/Impostor/Impostor/blob/c1c6311c87a1fddaf0d2d1987b551b8e090d2f5a/src/Impostor.Server/Net/Inner/Objects/InnerMeetingHud.cs#L152-L154
I will research it later

**Comment by Mr-VIT** [comment]
@realfaceitkantesilvestrs I believe it should be correct now

**Comment by realfaceitkantesilvestrs** [comment]
@Mr-VIT will test it out this evening

**Comment by realfaceitkantesilvestrs** [comment]
@Mr-VIT 
The votes are getting more accurate in general but there is one issue:

 - Before **Voting Session - 1** - Player **Andrew** was killed by **Night **
 - Andrew afterwards was recorded as voting

```
Voting Session - 1
Players alive: bloodz,Limur,titanium,MRJ,Odini,YGKEY,SwagY,Night,Hob
Limur voted bloodz
Andrew,bloodz,Odini,titanium,Hob,Night,MRJ,SwagY voted skip
Voting Session - 2
Players alive: bloodz,Limur,titanium,MRJ,Odini,YGKEY,SwagY,Night,Hob
Andrew,Hob,Night,MRJ voted skip
Limur,titanium,bloodz voted Hob
Odini,SwagY voted MRJ
Voting Session - 3
Players alive: bloodz,Limur,titanium,MRJ,Odini,YGKEY,SwagY,Night,Hob
SwagY,bloodz,MRJ,titanium voted Night
Limur,Odini,Andrew voted MRJ
Hob,Night voted skip
Voting Session - 4
Players alive: bloodz,Limur,titanium,MRJ,Odini,YGKEY,SwagY,Hob
titanium,SwagY,Limur,Andrew,Hob voted MRJ
bloodz,MRJ,Odini voted skip
Voting Session - 5
Players alive: bloodz,Limur,Odini,YGKEY,SwagY,Hob
Limur,Hob voted SwagY
SwagY,bloodz,Odini,Andrew voted Hob
```

**Comment by realfaceitkantesilvestrs** [comment]
@Mr-VIT 

I think we are almost there, here is one more issue:

Ok I have another issue `IPlayerExiledEvent` is called before the last `IPlayerVotedEvent` , in some scenarios if players submit their votes fast enough, multiple `IPlayerVotedEvent`'s are called after the `IPlayerExiledEvent`

```
[12/04/2020 03:42:26]    [JJJSAK]    [Info]    Meeting started
[12/04/2020 03:42:42]    [JJJSAK]    [Info]    Player McQuail voted for kirO
[12/04/2020 03:42:44]    [JJJSAK]    [Info]    Player jay voted for kirO
[12/04/2020 03:42:44]    [JJJSAK]    [Info]    Player eskay voted for kirO
[12/04/2020 03:42:47]    [JJJSAK]    [Info]    Player Cooch voted for kirO
[12/04/2020 03:42:47]    [JJJSAK]    [Info]    Player L voted for kirO
[12/04/2020 03:42:50]    [JJJSAK]    [Info]    Player Limur voted for kirO
[12/04/2020 03:42:53]    [JJJSAK]    [Info]    Player Dylan voted for kirO
[12/04/2020 03:43:33]    [JJJSAK]    [Info]    Players [McQuail,jay,eskay,Cooch,L,Limur,Dylan] voted out kirO <-- IPlayerExiledEvent
[12/04/2020 03:43:33]    [JJJSAK]    [Info]    kirO was voted off and marked with isDead = true <-- IPlayerExiledEvent
[12/04/2020 03:43:33]    [JJJSAK]    [Info]    Player kirO voted for kirO  <-- last IPlayerVotedEvent
[12/04/2020 03:43:43]    [JJJSAK]    [Info]    Meeting ended
```

**Comment by Mr-VIT** [comment]
host sends `RpcCalls.VotingComplete` before meeting HUD update data, as soon as received last vote

**Comment by realfaceitkantesilvestrs** [comment]
@AeonLucid  @Mr-VIT apart from the event order where player voted events in the last 0.5s come after meeting ended event I have had no issues with voting in the last 8 days testing with 30 people over 80 games.

Other than that issue - this seems to be ready to go.

**Comment by realfaceitkantesilvestrs** [comment]
@Mr-VIT @AeonLucid is it still an issue that this is not an async call?

**Comment by AeonLucid** [comment]
@js6pak decides

**Comment by js6pak** [comment]
Not like you can't reopen it yourself, no one is stopping you from rebasing from master and trying again

**Comment by Mr-VIT** [comment]
I can not reopen, I just don't have the button, but the repository authors do.

**Comment by js6pak** [comment]
I was pretty sure that you can open closed pull requests unless they are locked, my bad then. Make sure to rebase onto master and change this PR to master not dev.

File: src/Impostor.Server/Net/Inner/Objects/InnerMeetingHud.cs [hunk 4]
```diff
@@ -83,6 +85,37 @@ public override async ValueTask DeserializeAsync(IClientPlayer sender, IClientPl
                     if ((num & 1 << i) != 0)
                     {
                         _playerStates[i].Deserialize(reader);
+                        var playerState = _playerStates[i];
```
Comment (AeonLucid) [inline src/Impostor.Server/Net/Inner/Objects/InnerMeetingHud.cs:88]: Definitely needs to go in a separate method (@js6pak)

**Comment by JohnTranQUT** [comment]
I am unsure if this is an issue with the Impostor:2021.03.05 branch or with the merge with the Mr--VIT:event-vote
I've merge this pull in and have had this error come up quiet regularly.

```
 at Impostor.Server.Net.Inner.Objects.InnerPlayerInfo.Deserialize(IMessageReader reader) in src\Impostor.Server\Net\Inner\Objects\InnerPlayerInfo.cs:line 72
 at Impostor.Server.Net.Inner.Objects.InnerGameData.HandleRpcAsync(ClientPlayer sender, ClientPlayer target, RpcCalls call, IMessageReader reader) in src\Impostor.Server\Net\Inner\Objects\InnerGameData.cs:line 112
 at Impostor.Server.Net.State.Game.HandleGameDataAsync(IMessageReader parent, ClientPlayer sender, Boolean toPlayer) in src\Impostor.Server\Net\State\Game.Data.cs:line 213
 at Impostor.Server.Net.Client.HandleMessageAsync(IMessageReader reader, MessageType messageType) in src\Impostor.Server\Net\Client.cs:line 171
 at Impostor.Server.Net.Hazel.HazelConnection.ConnectionOnDataReceived(DataReceivedEventArgs e) in src\Impostor.Server\Net\Hazel\HazelConnection.cs:line 69
 at Impostor.Hazel.Connection.InvokeDataReceived(IMessageReader msg, MessageType messageType) in src\Impostor.Hazel\Connection.cs:line 184
  ```
Also getting this.

```
 System.IndexOutOfRangeException: Index was outside the bounds of the array.
 at Impostor.Hazel.MessageReader.ReadPackedUInt32() in src\Impostor.Hazel\MessageReader.cs:line 168
 at Impostor.Hazel.MessageReader.ReadPackedInt32() in src\Impostor.Hazel\MessageReader.cs:line 157
 at Impostor.Hazel.MessageReader.ReadString() in src\Impostor.Hazel\MessageReader.cs:line 139
 at Impostor.Server.Net.Inner.Objects.InnerPlayerInfo.Deserialize(IMessageReader reader) in src\Impostor.Server\Net\Inner\Objects\InnerPlayerInfo.cs:line 60
 at Impostor.Server.Net.Inner.Objects.InnerGameData.HandleRpcAsync(ClientPlayer sender, ClientPlayer target, RpcCalls call, IMessageReader reader) in src\Impostor.Server\Net\Inner\Objects\InnerGameData.cs:line 112
 at Impostor.Server.Net.State.Game.HandleGameDataAsync(IMessageReader parent, ClientPlayer sender, Boolean toPlayer) in src\Impostor.Server\Net\State\Game.Data.cs:line 213
 at Impostor.Server.Net.Client.HandleMessageAsync(IMessageReader reader, MessageType messageType) in src\Impostor.Server\Net\Client.cs:line 171
 at Impostor.Server.Net.Hazel.HazelConnection.ConnectionOnDataReceived(DataReceivedEventArgs e) in src\Impostor.Server\Net\Hazel\HazelConnection.cs:line 69
 at Impostor.Hazel.Connection.InvokeDataReceived(IMessageReader msg, MessageType messageType) in src\Impostor.Hazel\Connection.cs:line 184
```

Because I need the votes in my production environment , I am unable to make this error happen in my test/development environment with the sufficient number of scenarios for these errors to occur.

**Comment by kantesilvestrs** [comment]
@Mr-VIT @js6pak 

I pulled this branch and as I was calling a button I got this:

```text
[13:09:08 ERR] Invoking data received failed
System.NullReferenceException: Object reference not set to an instance of an object.
   at Impostor.Server.Net.Inner.Objects.InnerMeetingHud.<>c.<PopulateButtons>b__13_0(KeyValuePair`2 x) in /source/src/Impostor.Server/Net/Inner/Objects/InnerMeetingHud.cs:line 170
   at System.Linq.EnumerableSorter`2.ComputeKeys(TElement[] elements, Int32 count)
   at System.Linq.EnumerableSorter`1.ComputeMap(TElement[] elements, Int32 count)
   at System.Linq.EnumerableSorter`1.Sort(TElement[] elements, Int32 count)
   at System.Linq.OrderedEnumerable`1.GetEnumerator()+MoveNext()
   at System.Linq.Enumerable.SelectIPartitionIterator`2.PreallocatingToArray(Int32 count)
   at System.Linq.Enumerable.SelectIPartitionIterator`2.ToArray()
   at Impostor.Server.Net.Inner.Objects.InnerMeetingHud.PopulateButtons(Byte reporter) in /source/src/Impostor.Server/Net/Inner/Objects/InnerMeetingHud.cs:line 169
   at Impostor.Server.Net.Inner.Objects.InnerMeetingHud.DeserializeAsync(IClientPlayer sender, IClientPlayer target, IMessageReader reader, Boolean initialState) in /source/src/Impostor.Server/Net/Inner/Objects/InnerMeetingHud.cs:line 56
   at Impostor.Server.Net.State.Game.HandleGameDataAsync(IMessageReader parent, ClientPlayer sender, Boolean toPlayer) in /source/src/Impostor.Server/Net/State/Game.Data.cs:line 186
   at Impostor.Server.Net.Client.HandleMessageAsync(IMessageReader reader, MessageType messageType) in /source/src/Impostor.Server/Net/Client.cs:line 271
   at Impostor.Server.Net.Hazel.HazelConnection.ConnectionOnDataReceived(DataReceivedEventArgs e) in /source/src/Impostor.Server/Net/Hazel/HazelConnection.cs:line 69
   at Impostor.Hazel.Connection.InvokeDataReceived(IMessageReader msg, MessageType messageType) in /source/src/Impostor.Hazel/Connection.cs:line 192
```

Code:

```csharp
        private void PopulateButtons(byte reporter)
        {
            _playerStates = _gameNet.GameData!.Players
                .OrderBy(x => x.Value.Controller.NetId)
                .Select(x =>
                {
                    var area = new PlayerVoteArea(this, x.Key);
                    area.SetDead(x.Value.PlayerId == reporter, x.Value.Disconnected || x.Value.IsDead);
                    return area;
                })
                .ToArray();
        }
```

**Comment by JohnTranQUT** [comment]
@kantesilvestrs 

I'm wondering, did you merge into the main branch? For me it merged with conflicts. I had to restructure the code to match @js6pak 's main branch at 26th March 2021.
I works fine for me.
FYI, I've committed a pull request onto the involved in this pull request. That one should work for you.

**Comment by kantesilvestrs** [comment]
@JohnTranQUT  me and 6pak managed to make everything work today @Mr-VIT is going to update this PR and we'll be able to merge this in.

Last two commits that 6pak did to the master fixed the votes

**Comment by Mr-VIT** [comment]
The host player hold MeetingHud players list sorted by NetId. Server `InnerGameData._allPlayers` sorted by PlayerId. Previously used `_gameNet.GameData!.Players.OrderBy(x => x.Value.Controller?.NetId)` but now `NetId` is `MaxValue` for disconnected player, so `_allPlayers` can be out of order

https://github.com/Impostor/Impostor/blob/5f922f5f83c92628c04ce5d0f869404261b34dbc/src/Impostor.Server/Net/State/Game.Data.cs#L184

**Comment by js6pak** [comment]
Please fix all the warnings

**Comment by realfaceitkantesilvestrs** [comment]
@js6pak @Mr-VIT  are we good to go with this one?

**Comment by kantesilvestrs** [comment]
@Mr-VIT on the latest `airship` branch the last vote issue is becoming really common and starts to affect my results gathering alot more

```text
[04/07/2021 17:37:01]	[XGEOZT]	[Info]	Meeting started
[04/07/2021 17:37:51]	[XGEOZT]	[Info]	Player hw voted for rebecca
[04/07/2021 17:38:18]	[XGEOZT]	[Info]	Player rebecca voted for Monsterrr
[04/07/2021 17:38:29]	[XGEOZT]	[Info]	Player CraggY voted for rebecca
[04/07/2021 17:39:16]	[XGEOZT]	[Info]	Player Odini voted for rebecca
[04/07/2021 17:39:19]	[XGEOZT]	[Info]	Player Cheeze voted for rebecca
[04/07/2021 17:39:21]	[XGEOZT]	[Info]	Players [hw,CraggY,Odini,Cheeze] voted out rebecca
[04/07/2021 17:39:21]	[XGEOZT]	[Info]	Player Booyanach did not vote
[04/07/2021 17:39:21]	[XGEOZT]	[Info]	Meeting ended
```

That did not vote came at the same time as meeting ended and my game logger (not above) logged meeting ended before the `no vote` was logged, we need to figure out how to make this go one after the other, or maybe at least force.

Otherwise I need to artificially delay meeting ended call for a second same way I'm doing it for PLayerSpawnedEvent

**Comment by Mr-VIT** [comment]
@kantesilvestrs this a client-host issue, not a server. Wait a second after the meeting ends before finishing counting the votes.

**Comment by realfaceitkantesilvestrs** [comment]
@Mr-VIT there is issue with player order again, you can see `Player Booyanach voted for Booyanach` but he never voted for himself, and this happened in every meeting

```text
[04/11/2021 17:52:16]	[IABFLC]	[Info]	Meeting started
[04/11/2021 17:53:15]	[IABFLC]	[Info]	Player Cheeze voted for Booyanach
[04/11/2021 17:53:18]	[IABFLC]	[Info]	Player Odini skipped vote
[04/11/2021 17:53:19]	[IABFLC]	[Info]	Player Booyanach voted for Booyanach
[04/11/2021 17:53:21]	[IABFLC]	[Info]	Player Ox skipped vote
[04/11/2021 17:53:22]	[IABFLC]	[Info]	Player rebecca skipped vote
[04/11/2021 17:53:25]	[IABFLC]	[Info]	Player Andrew skipped vote
[04/11/2021 17:53:27]	[IABFLC]	[Info]	Player Hello voted for Booyanach
[04/11/2021 17:53:29]	[IABFLC]	[Info]	Player Luke skipped vote
[04/11/2021 17:53:29]	[IABFLC]	[Info]	Meeting ended
```

```text
[04/11/2021 17:58:09]	[IABFLC]	[Info]	Meeting started
[04/11/2021 17:58:14]	[IABFLC]	[Info]	Player Cheeze skipped vote
[04/11/2021 17:58:15]	[IABFLC]	[Info]	Player Andrew skipped vote
[04/11/2021 17:58:16]	[IABFLC]	[Info]	Player Booyanach voted for Booyanach
[04/11/2021 17:58:17]	[IABFLC]	[Info]	Player Ox skipped vote
[04/11/2021 17:58:18]	[IABFLC]	[Info]	Player rebecca skipped vote
[04/11/2021 17:58:26]	[IABFLC]	[Info]	Player Hello voted for Booyanach
[04/11/2021 17:58:29]	[IABFLC]	[Info]	Player Luke skipped vote
[04/11/2021 17:58:40]	[IABFLC]	[Info]	Player Odini skipped vote
[04/11/2021 17:58:40]	[IABFLC]	[Info]	Meeting ended
```

**Comment by realfaceitkantesilvestrs** [comment]
@js6pak  @Mr-VIT  what changes are we still waiting on this PR?

File: src/Impostor.Server/Net/State/Game.Data.cs [hunk 5]
```diff
@@ -381,6 +382,11 @@ private async ValueTask OnDestroyAsync(InnerNetObject netObj)
 
                 case InnerPlayerControl control:
                 {
+                    if (GameState != GameStates.Started)
+                    {
+                        GameNet.GameData?.RemovePlayer(control);
```
Comment (miniduikboot) [inline src/Impostor.Server/Net/State/Game.Data.cs:387]: I don't see yet how this code is related to the main subject of this PR

Comment (js6pak) [reply src/Impostor.Server/Net/State/Game.Data.cs:387]: Pretty sure it caused problems with player state array sorting, @Mr-VIT

File: src/Impostor.Api/Net/Inner/Objects/IInnerMeetingHud.cs [hunk 6]
```diff
@@ -1,6 +1,59 @@
...
+
+            /// <summary>
+            ///     Gets a value indicating whether the player is a dead.
+            /// </summary>
+            bool IsDead { get; }
+
+            /// <summary>
+            ///     Gets a value indicating whether the player started this meeting.
+            /// </summary>
+            bool DidReport { get; }
```
[the remark this answers is not in the document]
Comment (js6pak) [reply src/Impostor.Api/Net/Inner/Objects/IInnerMeetingHud.cs:24]: Kinda is, but its how it is on the client so eh

File: src/Impostor.Server/Net/State/Game.Data.cs (hunk 5 above)
Comment (realfaceitkantesilvestrs) [reply src/Impostor.Server/Net/State/Game.Data.cs:387]: @js6pak so does this resolve the issue where we had to sort by netid to get the proper player order?

Comment (Mr-VIT) [reply src/Impostor.Server/Net/State/Game.Data.cs:387]: this is necessary to match `GameData._allPlayers` with host.

**Comment by JohnTranQUT** [comment]
Shortly after a disconnect, it drops all players.
Happened during a meeting.
Please note this is the combination of this https://github.com/Impostor/Impostor/pull/181 and https://github.com/Impostor/Impostor/pull/395.
I've merge it into as: https://github.com/RankedAmongUs/Impostor/tree/2021.05.10v1/vote-task 

```
JBDEJH - Player Bearb (1) has left.
JBDEJH - Assigned Klonet (2) as new host.
JBDEJH - [1] has disconnected
Client 1 disconnecting, reason: Sent 6 pings that remote has not responded to.
Player eirik (3) tried to send DespawnFlag for 153 but was denied.
Player Robin (5) tried to send DespawnFlag for 153 but was denied.
Player Sandy (9) tried to send DespawnFlag for 153 but was denied.
Invoking data received failed
System.ArgumentOutOfRangeException: Index was out of range. Must be non-negative and less than the size of the collection. (Parameter 'index')
   at System.Collections.Generic.List`1.get_Item(Int32 index)
   at Impostor.Server.Net.Inner.Objects.InnerPlayerInfo.Deserialize(IMessageReader reader) in D:\Users\John\source\repos\Impostor\src\Impostor.Server\Net\Inner\Objects\InnerPlayerInfo.cs:line 83
   at Impostor.Server.Net.Inner.Objects.InnerGameData.DeserializeAsync(IClientPlayer sender, IClientPlayer target, IMessageReader reader, Boolean initialState) in Impostor\src\Impostor.Server\Net\Inner\Objects\InnerGameData.cs:line 93
   at Impostor.Server.Net.State.Game.HandleGameDataAsync(IMessageReader parent, ClientPlayer sender, Boolean toPlayer) in Impostor\src\Impostor.Server\Net\State\Game.Data.cs:line 90
   at Impostor.Server.Net.Client.HandleMessageAsync(IMessageReader reader, MessageType messageType) in Impostor\src\Impostor.Server\Net\Client.cs:line 176
   at Impostor.Server.Net.Hazel.HazelConnection.ConnectionOnDataReceived(DataReceivedEventArgs e) in Impostor\src\Impostor.Server\Net\Hazel\HazelConnection.cs:line 71
   at Impostor.Hazel.Connection.InvokeDataReceived(MessageReader msg, MessageType messageType) in Impostor\src\Impostor.Hazel\Hazel\Connection.cs:line 185
```

**Review by miniduikboot — COMMENTED** [review]
Looked through the code, haven't run it.

Some parts seem a bit suspicious, but it looks correct.

**Comment by JohnTranQUT** [comment]
@js6pak 

I can persistently make this error pop up.
What happens is that when a player leaves mid meeting, Invoking data fails.
**Invoking data received failed
 System.InvalidOperationException: Sequence contains no matching element**

Moments later, another invoking data error shows up.
**Invocation of event MeetingEndedEvent threw an exception.
 System.NullReferenceException: Object reference not set to an instance of an object.**

```batch
 [12:00:28 INF] Client 4 disconnecting, reason: The remote sent a disconnect request
 [12:00:28 INF] JWOTQY - Game ended by ImpostorByVote
 [12:00:28 WRN] Player Azn1 (3) tried to send DespawnFlag for 13 but was denied.
 [12:00:28 ERR] Invoking data received failed
 System.InvalidOperationException: Sequence contains no matching element
   at System.Linq.ThrowHelper.ThrowNoMatchException()
   at System.Linq.Enumerable.Single[TSource](IEnumerable`1 source, Func`2 predicate)
   at Impostor.Server.Net.Inner.Objects.InnerMeetingHud.DeserializeAsync(IClientPlayer sender, IClientPlayer target, IMessageReader reader, Boolean initialState) in Impostor\src\Impostor.Server\Net\Inner\Objects\InnerMeetingHud.cs:line 94
   at Impostor.Server.Net.State.Game.HandleGameDataAsync(IMessageReader parent, ClientPlayer sender, Boolean toPlayer) in Impostor\src\Impostor.Server\Net\State\Game.Data.cs:line 90
   at Impostor.Server.Net.Client.HandleMessageAsync(IMessageReader reader, MessageType messageType) in Impostor\src\Impostor.Server\Net\Client.cs:line 176
   at Impostor.Server.Net.Hazel.HazelConnection.ConnectionOnDataReceived(DataReceivedEventArgs e) in Impostor\src\Impostor.Server\Net\Hazel\HazelConnection.cs:line 71
   at Impostor.Hazel.Connection.InvokeDataReceived(MessageReader msg, MessageType messageType) in Impostor\src\Impostor.Hazel\Hazel\Connection.cs:line 185
 [12:00:28 INF] JWOTQY - Player Azn1 (2) has left.
 [12:00:28 INF] JWOTQY - Assigned Azn1 (3) as new host.
 [12:00:28 INF] Client 2 disconnecting, reason: Invoking data received failed
1|RAU Imp Node 1  | [12:00:51 DBG] Created game with code GMULDR.
1|RAU Imp Node 1  | [12:00:51 INF] GMULDR - Player AznDumbum (9) is joining.
1|RAU Imp Node 1  | [12:02:08 INF] GMULDR - Player AznDumbum (9) has left.
1|RAU Imp Node 1  | [12:02:08 DBG] Remove game with code GMULDR (GMULDR).
1|RAU Imp Node 1  | [12:02:08 INF] Destroying Game GMULDR
1|RAU Imp Node 1  | [12:02:08 INF] Client 9 disconnecting, reason: The remote sent a disconnect request
 [12:02:20 INF] JWOTQY - Azn1  abstained from voting.
 [12:02:20 INF] JWOTQY - Azn1 1 abstained from voting.
 [12:02:20 INF] JWOTQY - Azn1 2 abstained from voting.
 [12:02:20 ERR] Invocation of event MeetingEndedEvent threw an exception.
 System.NullReferenceException: Object reference not set to an instance of an object.
    at Impostor.Plugins.RankedAmongUs.Handlers.GameEventListener.OnMeetingEndedEvent(IMeetingEndedEvent e)
    at lambda_method24(Closure , Object , Object , IServiceProvider )
    at Impostor.Server.Events.Register.RegisteredEventListener.InvokeAsync(Object eventHandler, Object event, IServiceProvider provider) in D:\Users\John\source\repos\Impostor\src\Impostor.Server\Events\Register\RegisteredEventListener.cs:line 83
    at Impostor.Server.Events.Register.WrappedRegisteredEventListener.InvokeAsync(Object eventHandler, Object event, IServiceProvider provider) in D:\Users\John\source\repos\Impostor\src\Impostor.Server\Events\Register\WrappedRegisteredEventListener.cs:line 24
    at Impostor.Server.Events.EventManager.CallAsync[T](T event) in D:\Users\John\source\repos\Impostor\src\Impostor.Server\Events\EventManager.cs:line 90
 [12:02:56 INF] JWOTQY - Player Azn1 (3) has left.
 ```

**Comment by JohnTranQUT** [comment]
@js6pak https://github.com/Impostor/Impostor/pull/181/commits/50042fa7642ef63abed80bae9fad798957aa0b48 appears to have fixed the issues I was having. :)

**Pull request merged**